### PR TITLE
Fix issue 15523- extern(C++) support TLS (C++ 'thread_local') linkage

### PR DESCRIPTION
**Pull request by thewilsonator**

### CODE REVIEW

**Comment by thewilsonator** [comment]
Looks like we will have to do something special for OSX.

**Comment by ibuclaw** [comment]
This should be handled by the Target hooks, as the frontend has no knowledge of what kind of thread model has been configured (ie: Clang may well support native on OSX, but GCC still emulates TLS using a pthread library).

File: src/dmd/cppmangle.d [hunk 1]
```diff
@@ -995,8 +995,19 @@ private final class CppMangleVisitor : Visitor
...
+            if (global.params.targetOS == TargetOS.OSX)
+            {
+                // OSX emits TLS wrapper functions, shouldn't be hard to do.
+                // These start with _ZTW
+                d.error("C++ `thread_local` is unsupported on OSX");
+                fatal();
+            }
+            else if (global.params.cplusplus == CppStdRevision.cpp98)
+            {
+                d.error("C++ `thread_local` variables are not supported in C++98 compatibiity mode");
```
Comment (ibuclaw) [inline src/dmd/cppmangle.d:1007]: C++-98 supports `__thread`, or is that a gnu extension?

Comment (thewilsonator) [reply src/dmd/cppmangle.d:1007]: it is an extension, also support by clang.

File: src/dmd/cppmangle.d [hunk 2]
```diff
@@ -995,8 +995,19 @@ private final class CppMangleVisitor : Visitor
         // fake mangling for fields to fix https://issues.dlang.org/show_bug.cgi?id=16525
         if (!(d.storage_class & (STC.extern_ | STC.field | STC.gshared)))
         {
-            d.error("Internal Compiler Error: C++ static non-`__gshared` non-`extern` variables not supported");
-            fatal();
+            if (!target.cpp.threadLocalSupport())
```
Comment (ibuclaw) [inline src/dmd/cppmangle.d:998]: GCC supports `thread_local`, my comment was that you can't make any assumptions about how it is mangled in cppmangle. It is a red flag when you comment about OSX having a wrapper functions, because this is not true for GCC.

Comment (thewilsonator) [reply src/dmd/cppmangle.d:998]: > you can't make any assumptions about how it is mangled in cppmangle

What? Its not mangled at all. Thats why no mangling changes are necessary.

> OSX having a wrapper functions

The default environment for OSX is Apple clang which does use wrapper functions from the look of things (`_ZTW` symbols are expected which demangler.com tells me are wrapper functions). 

> because this is not true for GCC.

So what? This is DMD. That the comments don't match is irrelevant. LDC probably also works, but `target.d` is explicitly supposed to be complier specific.

Comment (ibuclaw) [reply src/dmd/cppmangle.d:998]: I've done some digging.  ABI isn't a concern, so I'll scratch this conversation, however TLS support is still impeded.

File: src/dmd/cppmangle.d [hunk 3]
```diff
@@ -995,8 +995,19 @@ private final class CppMangleVisitor : Visitor
...
+            {
+                // OSX emits TLS wrapper functions, shouldn't be hard to do.
+                // These start with _ZTW
+                d.error("C++ `thread_local` is unsupported for this target");
+                fatal();
```
Comment (ibuclaw) [inline src/dmd/cppmangle.d:1003]: Nope, *all* C++ compilers emit a TLS wrapper function which need to be called when interfacing with a variable such as:
```
extern(C++) extern int tlsVar;
```
So the only TLS variables that are safe to be `extern(C++)` are those that are defined on the D side - however there may be linker/runtime issues on the C++ if they try to access them (which is probably what you found out on OSX/clang, but this is conveniently hidden on GCC thanks to the kind of variables you're using to test not requiring any special construction.

Until we do this, we can't claim to have TLS support for C++ variables at all.

**Review by ibuclaw — CHANGES_REQUESTED** [review]

File: src/dmd/cppmangle.d (hunk 3 above)
Comment (ibuclaw) [reply src/dmd/cppmangle.d:1003]: For example:
```C++
#include <assert.h>
struct S
{
    S(int f);
    int field;
};
S::S(int f) : field(f) { }

extern thread_local S tls;

void fromCxx()
{
    assert(tls.field == 0xfeebdaed);
}
```
```D
extern(C++)
{
    struct S
    {
        this(int f);
        int field;
    }
    S tls;
    void fromCxx();
}

void main()
{
    tls = S(0xdeadbeef);
    fromCxx();
}

```
This won't link at all.

Comment (ibuclaw) [reply src/dmd/cppmangle.d:1003]: To flip the test around for variables defined on the C++ side.
```C++
#include <assert.h>
struct S
{
    S(unsigned f);
    unsigned field;
};
S::S(unsigned f) : field(f) { }

thread_local S tls = S(0xdeadbeef);

void fromCxx()
{
    assert(tls.field == 0xdeadbeef);
}
```
```D
extern(C++)
{
    struct S
    {
        this(uint f);
        uint field;
    }
    extern S tls;
    void fromCxx();
}

void main()
{
    assert(tls.field == 0xdeadbeef); // FAILS.
    fromCxx();
    assert(tls.field == 0xdeadbeef); // PASSES.
}
```

Comment (thewilsonator) [reply src/dmd/cppmangle.d:1003]: added that test case. lets see what fails

**Comment by ibuclaw** [comment]
So in summary - no, this error is not overzealous, we really do *not* support C++ TLS because:
1. We do not generate TLS initializer functions for variables D defines, referenced by C++.
2. We do not access `extern` TLS variables by their wrapper functions generated by C++.

For (1), we could get away with allowing trivial types, because we know on the C++ side, the implementation of the TLS wrapper function should look like:
```
ref T tlsWrapFoo()
{
  if (&tlsInitFoo !is null)
    tlsInitFoo();
  return foo;
} 
static if (__traits(isTriviallyInitialized, foo))
    @weak void tlsInitFoo();
else
    void tlsInitFoo();
```
But we can't assume that this is a given - not all targets support weak, and we can't know all reasons for a C++ compiler not emitting the init function as weak (otherwise OSX/clang would work with trivial types?) - and granted our approach to C++ support in D is to only reference symbols and generate *nothing*, that means ultimately we can only continue to reject locally defined C++ TLS variables.  This can be done with an error that isn't an ICE though.

For (2), we must *always* access externally defined C++ TLS variables by their wrapper functions, regardless of what type they are. We can generate the mangling for the wrapper functions easily enough, but the code generator needs to replace all C++ TLS variable accesses with a call to this wrapper function.

tl;dr
```
// Error because D can not and will not generate C++ code to initialize TLS variables.
// Our interaction with C++ code is to reference externally defined symbols only.
extern(C++) T var;

// We can support this, but must replace all accesses with a call to its wrapper function.
// cppmangle can generate the mangle of the extern _ZTW reference in the same
// way that we support C++ typeinfo.
extern(C++) extern T var;
```

**Comment by RazvanN7** [comment]
@thewilsonator What are your plans with this PR given @ibuclaw 's comments?
